app-2/src/main/java/tech/relaycorp/letro/awala/AwalaManager.kt

```diff
@@ -0,0 +1,235 @@
+import androidx.annotation.RawRes
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+import tech.relaycorp.awaladroid.Awala
+import tech.relaycorp.awaladroid.GatewayBindingException
+import tech.relaycorp.awaladroid.GatewayClient
+import tech.relaycorp.awaladroid.endpoint.FirstPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.InvalidThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.PublicThirdPartyEndpoint
+import tech.relaycorp.awaladroid.endpoint.ThirdPartyEndpoint
+import tech.relaycorp.awaladroid.messaging.OutgoingMessage
+import tech.relaycorp.letro.R
+import tech.relaycorp.letro.awala.message.Message
+import tech.relaycorp.letro.awala.message.MessageType
+import tech.relaycorp.letro.ui.navigation.Route
+import tech.relaycorp.letro.utils.awala.loadNonNullFirstPartyEndpoint
+import tech.relaycorp.letro.utils.awala.loadNonNullThirdPartyEndpoint
+import javax.inject.Inject
+
+interface AwalaManager {
+    val messages: Flow<Message>
+    suspend fun sendMessage(message: Message)
+    suspend fun isAwalaInstalled(currentScreen: Route): Boolean
+}
+
+class AwalaManagerImpl @Inject constructor(
+    private val awalaRepository: AwalaRepository,
+    @ApplicationContext private val context: Context,
+): AwalaManager {
+
+    private val awalaScope = CoroutineScope(Dispatchers.IO)
+
+    private val _messages = MutableSharedFlow<Message>()
+    override val messages: Flow<Message>
+        get() = _messages
+
+    @Volatile
+    private var isAwalaSetUp = false
+    private var awalaSetupJob: Job? = null
+
+    @Volatile
+    private var isAwalaInstalledOnDevice: Boolean? = null
+
+    @Volatile
+    private var isReceivingMessages = false
+
+    init {
+        awalaSetupJob = awalaScope.launch {
+            Awala.setUp(context)
+            checkIfAwalaAppInstalled()
+            if (isAwalaInstalledOnDevice == true) {
+                configureAwala()
+            }
+            isAwalaSetUp = true
+            awalaSetupJob = null
+        }
+    }
+
+    override suspend fun sendMessage(message: Message) {
+        awalaSetupJob?.join()
+        val firstPartyEndpointNodeId = awalaRepository.getServerFirstPartyEndpointNodeId()
+            ?: registerFirstPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        val thirdPartyEndpointNodeId = awalaRepository.getServerThirdPartyEndpointNodeId()
+            ?: importServerThirdPartyEndpointIfNeeded()?.nodeId
+            ?: ""
+
+        if (firstPartyEndpointNodeId.isEmpty() || thirdPartyEndpointNodeId.isEmpty()) {
+            // TODO: log error
+            Log.i(TAG, "sendMessage(). some of ids is empty: $firstPartyEndpointNodeId ; $thirdPartyEndpointNodeId")
+            return
```

Comment:
I think this should crash the app. It'd be a bug.

Reply:
Done